### PR TITLE
Deprecate alias_method_chain in favour of Module#prepend

### PR DESCRIPTION
**Pull request by kirs**
…as discussed #19413
@rafaelfranca

### CODE REVIEW

**Comment by rafaelfranca** [comment]
We are still using it internally right? Lets remove all the calls before deprecating.

File: activesupport/lib/active_support/core_ext/module/aliasing.rb [hunk 1]
```diff
@@ -21,6 +21,8 @@ class Module
   #
   # so you can safely chain foo, foo?, foo! and/or foo= with the same feature.
   def alias_method_chain(target, feature)
+    ActiveSupport::Deprecation.warn("alias_method_chain is deprecated. Please, use Module#prepend instead.")
```
Comment (kaspth) [inline activesupport/lib/active_support/core_ext/module/aliasing.rb]: `alias_method_chain` is usually used to extend an existing method and call the original. This deprecation method leaves the last part implicit. Let's mention you can access the original method using super.

**Comment by kirs** [comment]
@rafaelfranca done.

Now there still 2 places where `alias_method_chain` is called:

* `Range` core_ext. We can't switch to Module#prepend there yet because of bug in MRI (https://bugs.ruby-lang.org/issues/10847)
* in `alias_method_chain` tests

**Comment by rafaelfranca** [comment]
@kirs in Rage core_ext lets remove the usage of `alias_method_chain` with its implementation, so we don't get deprecation warnings in our test, and add a TODO note to remove it as soon the bug is fixed.

**Comment by kirs** [comment]
@rafaelfranca done!

**Comment by dhh** [comment]
I don't see how this is a replacement? Prepend operates on the whole module, requires the method that's beind prepended to call super. If we would have gotten the proposed features of before/after/around, that would have been a straight up replacement. This isn't.

**Comment by rafaelfranca** [comment]
@dhh the method that call `super` is the method that is extending, the original method don't need to take care of it. Could you show an example where `prepend` doesn't work?

**Comment by rafaelfranca** [comment]
When we proposed `prepend` to Ruby core was with the idea of being a complete replacement for `alias_method_chain`. I may be missing a case where `prepend` doesn't work, but I always could replace `alias_method_chain` with `prepend` without touching the original method.

**Comment by dhh** [comment]
Ah, okay, maybe I missed that. I’ll have another look.

**Comment by rafaelfranca** [comment]
Right. Let me know if you find a case, so we can remove the deprecation.

File: activesupport/CHANGELOG.md [hunk 2]
```diff
@@ -1,3 +1,7 @@
+*   Deprecate `alias_method_chain` in favour of `Module#prepend` introduced in Ruby 2.0
```
Comment (kaspth) [inline activesupport/CHANGELOG.md]: @kirs Rails tries to stick to american english, so next time prefer favor over favour etc. :smile:

**Pull request merged**